propose fix for CORS issues with missing crossorigin tag

PhilETaylor opened the pull request:
This fixes https://github.com/symfony/webpack-encore-bundle/issues/55 and comes with the unit tests too ;-) 

updated:

This PR also allows the value for crossorigin attribute to be defined in the configuration yaml

PhilETaylor committed:
propose fix for CORS issues with missing crossorigin tag

https://github.com/symfony/webpack-encore-bundle/issues/55

PhilETaylor committed:
fix tests for  crossorigin="anonymous"

PhilETaylor committed:
fix tests

Lyrkan commented:
Hi @PhilETaylor,

Indeed, that's an issue in a cross-domain context (which is not always the case).

I'm not against adding `crossorigin="anonymous"` but are we sure that other people with different setups won't want/need to use other (or no) values for that attribute?

I would 100% agree with using `anonymous` as the default value if there was also a way to override it, either by adding something to `enableIntegrityHashes` or directly when calling `{{ encore_entry_script_tags }}` (kinda related: https://github.com/symfony/webpack-encore-bundle/issues/10)

PhilETaylor commented:
> Indeed, that's an issue in a cross-domain context (which is not always the case).

Its the case 100% of the time if you are using a CDN to serve your assets :) 

I dont have any strong opinion either way, however when I upgraded and deployed to production (which uses a cdn) it broke my site.. hence I had to work out why, document it, and find a solution.. which I have proposed. 

The current situation is that anyone that follow what I did, updates, enables integrity, and then deploys with assets from a cdn will break their site...

PhilETaylor commented:
there are only two options really for the crossorigin attribute. https://developer.mozilla.org/en-US/docs/Web/HTML/CORS_settings_attributes

`anonymous` and `use-credentials`

I've never seen anyone use the latter... but at the moment this feature is BROKEN because deploying assets on a CDN with the integrity attribute and no crossorigin attribute **WILL BREAK THE PRODUCTION SITE!**  (break == asset loading will be blocked by Google Chrome)

I would say a quick fix would be the hard coding of crossorigin as proposed by this PR and then further discussion on making it configurable can be had later, it seems #10 has been open since last year with no progress and you cant expect to leave this feature broken for the majority while someone makes that decision

PhilETaylor committed:
Allow value for crossorigin attribute to be configurable, defaults to anonymous

PhilETaylor commented:
ok with 768a820 I have now made the value of crossorigin attribute configurable  :-) 

if none is set in the config, then it still defaults to anonymous

PLEASE TEST!

PhilETaylor committed:
update documentation

PhilETaylor commented:
flex recipe also updated here https://github.com/symfony/recipes/pull/562

stof commented on src/DependencyInjection/Configuration.php:
```diff
@@ -34,6 +34,9 @@ public function getConfigTreeBuilder()
                     ->isRequired()
                     ->info('The path where Encore is building the assets - i.e. Encore.setOutputPath()')
                 ->end()
+                ->scalarNode('crossorigin')
```
as there is only 2 valid values, I suggest using an `enumNode` instead

PhilETaylor replied:
Done.

stof commented on src/DependencyInjection/WebpackEncoreExtension.php:
```diff
@@ -51,6 +51,9 @@ public function load(array $configs, ContainerBuilder $container)
         $container->setAlias(EntrypointLookupInterface::class, new Alias($this->getEntrypointServiceId('_default')));
+
+        $container->getDefinition('webpack_encore.tag_renderer')
+            ->replaceArgument(2, $config['crossorigin']??'anonymous');
```
The default value should be moved to the Configuration class. This way, tooling can now about it.

PhilETaylor replied:
Done.

stof commented on src/Resources/config/services.xml:
```diff
@@ -16,6 +16,7 @@
         <service id="webpack_encore.tag_renderer" class="Symfony\WebpackEncoreBundle\Asset\TagRenderer">
             <argument type="service" id="webpack_encore.entrypoint_lookup_collection" />
             <argument type="service" id="assets.packages" />
+            <argument type="string" id="webpack_encore.crossorigin" />
```
`id` is not a valid attribute for `<argument type="string">` (and we generally omit `type="string"` entirely as that's the default behavior too)

PhilETaylor replied:
I think I have done this right now, just `<attribute>anonymous</attribute>` ???

PhilETaylor committed:
remove additional spaces

PhilETaylor commented:
@stof Thanks for the comments... coding on the sofa on laptop is not my best place to code :) also still learning some of the deep symfony stuff :) ... "it works" is not enough :) I'll go grab a coffee and update this PR with your comments - thanks for the feedback

PhilETaylor committed:
refine as per feedback

PhilETaylor commented:
@stof c304692 has changes based on your feedback, only thing Im not sure is if the services xml should be 

`<argument/>`

or the repetition of the default value like 

`<argument>anonymous</argument>`

I could not find any examples to learn from on this :)

PhilETaylor committed:
<argument /><!-- crossorigin -->

PhilETaylor commented:
ok I have settled on 

`<argument /><!-- crossorigin -->`

as I found that example in symfony/security-bundle like that :)

PhilETaylor commented:
This is ready for testing/feedback again now :)

Lyrkan commented on src/DependencyInjection/Configuration.php:
```diff
@@ -34,6 +34,11 @@ public function getConfigTreeBuilder()
                 ->end()
+                ->enumNode('crossorigin')
+                    ->defaultValue('anonymous')
+                    ->values(['anonymous', 'use-credentials'])
```
I'm not sure if this would be useful or not but should we also handle a `false` case in which the attribute is not added?

PhilETaylor replied:
Correct. It is possible to have the assets on the same domain name as your site, and therefore integrity can be set without a crossorigin attribute. I have now made changes to support this by allowing the default/false to not show the attribute.

Lyrkan commented on src/Asset/TagRenderer.php:
```diff
@@ -18,9 +18,12 @@ final class TagRenderer
         $entrypointLookupCollection,
-        Packages $packages
+        Packages $packages,
+        string $crossorigin
```
Maybe it would be better to have a `$defaultAttributes` array here instead? in case we want to be able to configure more of them later?

PhilETaylor replied:
Yes. Done.

Lyrkan commented on README.md:
```diff
@@ -24,7 +24,10 @@ webpack_encore:
     # output_path: false
-
+    
+    # if using Encore.enableIntegrityHashes() specify the crossorigin attribute value (default: anonymous)
```
Also, do we really want to only add it when the `integrity` attribute is also present? Couldn't it make sense to also need the `crossorigin` attribute for other reasons?

PhilETaylor replied:
Correct. It is possible to have the assets on the same domain name as your site, and therefore integrity can be set without a crossorigin attribute. I have now made changes to support this by allowing the default/false to not show the attribute.

PhilETaylor committed:
Allow for no attribute by default, and allow it to be configurable and future proofed array

PhilETaylor committed:
Fix unit tests

PhilETaylor committed:
correct documentation of configuration

PhilETaylor commented:
@Lyrkan all your feedback has now been applied. Ready for testing/feedback/merging again now

PhilETaylor commented:
flex recipe also updated here symfony/recipes#562

stof commented on src/Asset/TagRenderer.php:
```diff
@@ -54,6 +58,10 @@ public function renderWebpackScriptTags(string $entryName, string $packageName =
                 $attributes['integrity'] = $integrityHashes[$filename];
             }
 
+            if (isset($this->defaultAttributes['crossorigin']) && false !== $this->defaultAttributes['crossorigin']) {
```
this does not make sense. If you take an array `$defaultAttributes` in the renderer, you should merge them in attributes, not require implementing support for them one by one.

PhilETaylor replied:
ok thats probably beyond my symfony skill level - sorry. Happy for you to propose the code, show me an example, or take over this?

stof replied:
Well, instead of `$attributes = [ 'src' => ... ]` at the beginning, make it like that:

```php
$attributes = $this->defaultAttributes;
$attributes['src'] = ...
```

there is nothing specific to Symfony here.

stof replied:
note that this snippet assumes that `$this->defaultAttributes` contains only attributes which should be applied, not keys with `false` values when omitting them.

PhilETaylor replied:
ah I thought you meant something with an options resolver :) 

ok I'll give it another go :)

PhilETaylor replied:
ok committed a change to incorporate your feedback. ok?

stof commented on src/DependencyInjection/WebpackEncoreExtension.php:
```diff
@@ -51,6 +51,11 @@ public function load(array $configs, ContainerBuilder $container)
+
+        $container->getDefinition('webpack_encore.tag_renderer')
+            ->replaceArgument(2, [
+                'crossorigin' => $config['crossorigin'],
```
I would rather filter out `false` value here at compile time rather than filtering them at runtime.

PhilETaylor replied:
I think I have implemented this like you recommend. ok?

PhilETaylor committed:
refactor based on @stof feedback

PhilETaylor committed:
Apply codestyle from php-cs-fixer fix --config=.php_cs.dist

stof commented on src/Resources/config/services.xml:
```diff
@@ -16,6 +16,9 @@
             <argument type="service" id="webpack_encore.entrypoint_lookup_collection" />
             <argument type="service" id="assets.packages" />
+            <argument type="collection">
+                <argument key="crossorigin" />
```
this one should be removed. The default config should be an empty array (so `<argument type="collection" />`)

PhilETaylor replied:
Done.

PhilETaylor committed:
tidy up service xml to remove keyed collection

Lyrkan reviewed (commented):
Looks good to me :)

Maybe we're just missing some test cases for the default and explicit `false` values?

PhilETaylor committed:
revert the default test so that it tests with no attribute array

PhilETaylor commented:
In 7eeb941 I have reverted (and renamed) one of the tests so that it renders with no configured attributes, this is a quick win and covers the [TagRenderer](https://github.com/symfony/webpack-encore-bundle/blob/master/src/Asset/TagRenderer.php) class only, I guess what is really needed is some coverage of the [WebpackEncoreExtension](https://github.com/symfony/webpack-encore-bundle/blob/master/src/DependencyInjection/WebpackEncoreExtension.php) final class to check if the configuration attributes resolve correctly but there was no existing test and im not in the mood for that challenge today, sorry :) :)

I have also notices that LOADS of DocBlocks are missing on methods in this bundle - any ideas if that is on purpose or is that another quick win we could make using phpStorm to generate these missing DocBlocks?

Lyrkan commented:
>  I guess what is really needed is some coverage of the WebpackEncoreExtension final class to check if the configuration attributes resolve correctly but there was no existing test

The extension is tested by the [`IntegrationTest`](https://github.com/symfony/webpack-encore-bundle/blob/master/tests/IntegrationTest.php):

https://github.com/symfony/webpack-encore-bundle/blob/fc18456ed44737995f912c2084505fb67d040e11/tests/IntegrationTest.php#L147-L153

> I have also notices that LOADS of DocBlocks are missing on methods in this bundle - any ideas if that is on purpose or is that another quick win we could make using phpStorm to generate these missing DocBlocks?

Do you have an example of what you'd add? Most methods are already typed (params and return) so it wouldn't be really useful to add the same thing using PHPDoc.

Anyway, this should probably be in a separate PR to avoid mixing those changes with the ones here :)

PhilETaylor commented:
I'm just not used to seeing code without the additional PHPDoc for every method :-) I'll ignore that for now, phpStorm is great at reading the typehints so doesnt need the documentation to build its index anyway

I have added the default crossorigin = false in the IntegrationTest, the tests pass with it set to false, or not there at all. 

I guess this is ready for merging ?

PhilETaylor committed:
add default attribute set to false in tests

weaverryan committed:
making arg optional for BC

weaverryan committed:
small docs tweak

weaverryan committed:
Merge branch 'master' into patch-1

weaverryan commented:
Thanks for researching and fixing this Phil - excellent PR!

PhilETaylor commented:
🎉
